internal/backend/db/dbgorm/deployments.go

```diff
@@ -53,20 +54,43 @@ func (gdb *gormdb) deleteDeploymentsAndDependents(ctx context.Context, depIDs []
 	return db.Delete(&scheme.Deployment{}, "deployment_id IN ?", depIDs).Error
 }
 
-func (gdb *gormdb) updateDeploymentState(ctx context.Context, deploymentID sdktypes.UUID, state sdktypes.DeploymentState) error {
-	d := &scheme.Deployment{DeploymentID: deploymentID}
-
-	return gdb.transaction(ctx, func(tx *tx) error {
+func (gdb *gormdb) updateDeploymentState(
+	ctx context.Context,
+	deploymentID sdktypes.UUID,
+	state sdktypes.DeploymentState,
+) (sdktypes.DeploymentState, error) {
+	// NOTES:
+	// - we want to return the old state of the deployment before it was updated.
+	// - using RETURNING clause in UPDATE will return the new value, not the old one
+	// - in postgres it's possible to use `UPDATE tbl t1 ... FROM tbl t2.. RETURNING t2.column`,
+	//   e.g. effectively returning the old value before being updated). Unfortunately it won't work
+	//   in SQLite, since RETURNING could use only canonical table name, and not the alias.
+	// - that's why we are using two queries: one to get the old value, and the other to update it.
+
+	d := scheme.Deployment{DeploymentID: deploymentID}
+	var oldState int32 = 0
+
+	if err := gdb.transaction(ctx, func(tx *tx) error {
 		if err := tx.isCtxUserEntity(ctx, deploymentID); err != nil {
 			return err
 		}
 
-		result := tx.db.Model(d).Updates(map[string]any{"state": int32(state.ToProto()), "updated_at": time.Now()})
-		if result.Error != nil {
-			return result.Error
+		if err := tx.db.Model(&d).Select("state").First(&oldState).Error; err != nil {
+			return err
+		}
+
+		if err := tx.db.Model(&d).Update("state", int32(state.ToProto())).Error; err != nil {
+			return err
 		}
 		return nil
-	})
+	}); err != nil {
+		return sdktypes.DeploymentStateUnspecified, err
+	}
+	state, err := sdktypes.DeploymentStateFromProto(deploymentsv1.DeploymentState(oldState))
+	if err != nil {
```

Comment:
so what about the transaction in this case ?
it already happened, 
what's the system state now ?

Reply:
I think if you might have an invalid state in the db (this is what you check here)
then you should fail the transaction instead of committing it and then return error on the old state

Reply:
I could change it to kittehs.must1.
we cannot have invalid state in DB, it's converted from sdktypes.DeploymentState which defined well and cannot be some unknown value

Reply:
changed to kittehs.Must1